src/report/ReportPoll.ts

```diff
@@ -0,0 +1,132 @@
+     * @param save An abstract function to persist where we got to in report reading
+     */
+    constructor(
+        private mjolnir: Mjolnir,
+        private manager: ReportManager,
+    ) { }
+
+    private schedulePoll() {
+        if (this.timeout === null) {
+            /*
+             * Important that we use `setTimeout` here, not `setInterval`,
+             * because if there's networking problems and `getAbuseReports`
+             * hangs for longer thank the interval, it could cause a stampede
+             * of requests when networking problems resolve
+             */
+            this.timeout = setTimeout(
+                this.getAbuseReports.bind(this),
+                60_000 // a minute in milliseconds
+            );
+        } else {
+            throw new InvalidStateError("poll already scheduled");
+        }
+    }
+
+    private async getAbuseReports() {
+        this.timeout = null;
+
+        let response_: {
+            event_reports: { room_id: string, event_id: string, sender: string, reason: string }[],
+            next_token: number | undefined
+        } | undefined = undefined;
+        try {
+            response_ = await this.mjolnir.client.doRequest(
+                "GET",
+                "/_synapse/admin/v1/event_reports",
+                { from: this.from.toString() }
+            );
+        } catch (ex) {
+            await this.mjolnir.logMessage(LogLevel.ERROR, "getAbuseReports", `failed to poll events: ${ex}`);
+            this.schedulePoll();
+            return;
+        }
+
+        const response = response_!;
+        for (let report of response.event_reports) {
+            let event: any; // `any` because `handleServerAbuseReport` uses `any`
+            try {
+                event = (await this.mjolnir.client.doRequest(
+                    "GET",
+                    `/_synapse/admin/v1/rooms/${report.room_id}/context/${report.event_id}?limit=1`
+                )).event;
+            } catch (ex) {
+                continue;
+            }
+
+            await this.manager.handleServerAbuseReport({
+                roomId: report.room_id,
+                reporterId: report.sender,
+                event: event,
+                reason: report.reason,
+            });
+        }
+
+        /*
+         * This API endpoint returns an opaque `next_token` number that we
+         * need to give back to subsequent requests for pagination, so here we
+         * save it in account data
+         */
+        if (response.next_token !== undefined) {
+            this.from = response.next_token;
+            try {
+               await this.mjolnir.client.setAccountData(REPORT_POLL_EVENT_TYPE, { from: response.next_token });
+            } catch (ex) {
+                await this.mjolnir.logMessage(LogLevel.ERROR, "getAbuseReports", `failed to update progress: ${ex}`);
+            }
+        }
+
+        this.schedulePoll();
```

Comment:
I'm generally averse to wrapping large chunks of code in catchalls; do we expect anything in that loop to be able to error without some sort of much bigger problem?

Reply:
I don't expect anything, but that's the thing with servers, sometimes the unexpected happen.

I guess the big question is do we gain something from dropping the loop in case of exception vs. do we gain from not dropping it in case of exception?

My take on it is:

# Drop it

+ Maybe we have a higher chance of realizing that something is broken, because breakages will be larger? No guarantee, though, as we don't have a good way to be informed in case of breakage.
- In case of breakage, we need to restart Mjölnir. Which could be at 2am.

# Keep it alive

+ In line with what we try to do with the rest of Mjölnir, we attempt to maintain functionality even in case of minor issue.
+ More (hopefully relevant) errors will show up in the logs.
- Higher risk that we may not realize that there is a bug at all, if we don't look at all.

So I'd go with "keep the loop alive" :)